pkg/api/kubegraph/analysis/hpa.go

```diff
@@ -0,0 +1,155 @@
+	osgraph "github.com/openshift/origin/pkg/api/graph"
+	"github.com/openshift/origin/pkg/api/kubegraph"
+	kubenodes "github.com/openshift/origin/pkg/api/kubegraph/nodes"
+	deploygraph "github.com/openshift/origin/pkg/deploy/graph"
+	deploynodes "github.com/openshift/origin/pkg/deploy/graph/nodes"
+)
+
+const (
+	HPAMissingCPUTargetError      = "HPAMissingCPUTarget"
+	HPAOverlappingScaleRefWarning = "HPAOverlappingScaleRef"
+)
+
+// FindHPASpecsMissingCPUTargets scans the graph in search of HorizontalPodAutoscalers that are missing a CPU utilization target.
+// As of right now, the only metric that HPAs can use to scale pods is the CPU utilization, so if a HPA is missing this target it
+// is effectively useless.
+func FindHPASpecsMissingCPUTargets(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	for _, uncastNode := range graph.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		hpa, ok := node.Object().(*extensions.HorizontalPodAutoscaler)
+		if !ok {
+			continue
+		}
+
+		if hpa.Spec.CPUUtilization == nil {
+			markers = append(markers, osgraph.Marker{
+				Node:     node,
+				Severity: osgraph.ErrorSeverity,
+				Key:      HPAMissingCPUTargetError,
+				Message:  fmt.Sprintf("%s is missing a CPU utilization target", namer.ResourceName(node)),
+			})
+		}
+	}
+
+	return markers
+}
+
+// FindOverlappingHPAs scans the graph in search of HorizontalPodAutoscalers that are attempting to scale the same set of pods.
+// This can occur in two ways:
+//   - 1. label selectors for two ReplicationControllers/DeploymentConfigs/etc overlap
+//   - 2. multiple HorizontalPodAutoscalers are attempting to scale the same ReplicationController/DeploymentConfig/etc
+// Case 1 is handled by deconflicting the area of influence of ReplicationControllers/DeploymentConfigs/etc, and therefore we
+// can assume that it will be handled before this step. Therefore, we are only concerned with finding HPAs that are trying to
+// scale the same resources.
+func FindOverlappingHPAs(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := &markerPool{}
+
+	nodeFilterFunction := osgraph.NodesOfKind(
+		kubenodes.HorizontalPodAutoscalerNodeKind,
+		kubenodes.ReplicationControllerNodeKind,
+		deploynodes.DeploymentConfigNodeKind,
+	)
+	edgeFilterFunction := osgraph.EdgesOfKind(
+		kubegraph.ScalingReplicationControllerEdgeKind,
+		kubegraph.ScalingDeploymentConfigEdgeKind,
+		deploygraph.DeploymentEdgeKind,
+	)
+
+	hpaSubGraph := graph.Subgraph(nodeFilterFunction, edgeFilterFunction)
+
+	for _, uncastNode := range hpaSubGraph.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		generateOverlappingHPAMarkers(namer.ResourceName(node), []graphapi.Node{uncastNode}, markers, uncastNode, hpaSubGraph, namer)
+	}
+
+	return markers.Markers
+}
+
+type markerPool struct {
```

Comment:
There is https://github.com/openshift/origin/blob/d64f7a178b5cf4f18cfe84fdc22500fc3e0011aa/pkg/api/graph/interfaces.go#L38